<a href="https://colab.research.google.com/github/Faraaztqureshi/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/Faraaz_Qureshi_006_DS10_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

Your assignment is to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

Your goals:

1. Load and clean the data (or determine the best method to drop observations when running tests)
2. Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01
3. Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01
4. Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 (i.e. there may not be much of a difference)

Note that this data will involve *2 sample* t-tests, because you're comparing averages across two groups (republicans and democrats) rather than a single group against a null hypothesis.

Stretch goals:

1. Refactor your code into functions so it's easy to rerun with arbitrary variables
2. Apply hypothesis testing to your personal project data (for the purposes of this notebook you can type a summary of the hypothesis you formed and tested)

In [0]:
## lets load the relevant libraries
import pandas as pd
import numpy as np
from scipy.stats import ttest_ind
from scipy.stats import ttest_1samp


### **Data Pre-Processing**

In [2]:
## now to get the dataset
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2019-11-04 19:45:37--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.03s   

2019-11-04 19:45:38 (636 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [3]:
## import the dataset using pandas
## I'll use the headers provided in class
## this code also deals with ? and makes them NaN's
column_headers = ['party','handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

df = pd.read_csv('house-votes-84.data', 
                 header=None, 
                 names=column_headers,
                 na_values="?")

df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n          NaN
2    democrat                 NaN             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
## since yay and nay are categorical we are going to assign them a value 
## 1 means yay and 0 means nay
## this way means that are closer to one means more support
## and means closer to 0 have little support
df = df.replace({'y': 1, 'n': 0})

In [5]:
## check if it worked
df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [0]:
## now to seperate into republicans and democrats
reps = df[df['party'] == 'republican']
dems = df[df['party'] == 'democrat']

In [12]:
## check to see how many of republicans should be 168
reps.shape

(168, 17)

In [13]:
## check to see how many democrats should be 267
dems.shape

(267, 17)

### Using hypothesis testing, find an issue that democrats support more than republicans with p < 0.01

In [16]:
## here are the options 
list(df.columns)

['party',
 'handicapped-infants',
 'water-project',
 'budget',
 'physician-fee-freeze',
 'el-salvador-aid',
 'religious-groups',
 'anti-satellite-ban',
 'aid-to-contras',
 'mx-missile',
 'immigration',
 'synfuels',
 'education',
 'right-to-sue',
 'crime',
 'duty-free',
 'south-africa']

In [17]:
## Lets try right-to-sue 
## lets check the mean for democrats
dems['right-to-sue'].mean()

0.2896825396825397

In [18]:
reps['right-to-sue'].mean()

0.8607594936708861

In [19]:
## by examining the means it seems like there is a pretty strong difference 
## I am looking right now for something that democrats support

dems.describe()


handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [20]:
reps.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [22]:
## by doing .describe() I can more easily compare
## looks like dems support budget bill (mean .888)
## looks like reps do not support budget bill (mean .134)
## t-test time
ttest_ind(reps['budget'], dems['budget'], nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

Null Hypothesis: There will be no difference between Republican and Democrat support for the budget bill.

Due to a p value of .0000000000000000000000000000000000000000000000000000000000000000000000000207 we can reject the null hypothesis and conclude that it is something different.

### Using hypothesis testing, find an issue that republicans support more than democrats with p < 0.01

In [24]:
## now I can go back to the right to sue issue
## reps (mean .860) dems (mean .289)
## I'm writing out the zeros to show I understand the scientific notation

ttest_ind(reps['right-to-sue'], dems['right-to-sue'], nan_policy='omit')

Ttest_indResult(statistic=13.51064251060933, pvalue=1.2278581709672758e-34)

Null Hypothesis:  There will be no difference between Republican and Democrat support for the right-to-sue bill.  Due to p value of .0000000000000000000000000000000122 we can reject the null hypothesis and conclude it is something different.  

### Using hypothesis testing, find an issue where the difference between republicans and democrats has p > 0.1 

In [25]:
## to answer this question I am going to back to the means and find some that are close together
reps.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           165.000000     148.000000  ...  156.000000    146.000000
mean              0.187879       0.506757  ...    0.089744      0.657534
std               0.391804       0.501652  ...    0.286735      0.476168
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      0.000000
50%               0.000000       1.000000  ...    0.000000      1.000000
75%               0.000000       1.000000  ...    0.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [26]:
dems.describe()

handicapped-infants  water-project  ...   duty-free  south-africa
count           258.000000     239.000000  ...  251.000000    185.000000
mean              0.604651       0.502092  ...    0.637450      0.935135
std               0.489876       0.501045  ...    0.481697      0.246956
min               0.000000       0.000000  ...    0.000000      0.000000
25%               0.000000       0.000000  ...    0.000000      1.000000
50%               1.000000       1.000000  ...    1.000000      1.000000
75%               1.000000       1.000000  ...    1.000000      1.000000
max               1.000000       1.000000  ...    1.000000      1.000000

[8 rows x 16 columns]

In [27]:
## looking at the water-project bill the means are very similar so lets try that
## now my null hypothesis is that there is no difference between republicans and democrats on the water project bill
ttest_ind(reps['water-project'], dems['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

Null Hypothesis:  There will be no difference in the level of support for the water-project bill between Republicans and Democrats.  Due to a p value of .929 we fail to reject the null hypothesis.

In [28]:
## I came back to the assignment after looking over some notes
## Above I changed the null hypothesis after watching the video I realize we
## only change the null hypothesis when doing a 1-sample test
ttest_ind(reps['water-project'], dems['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

**Null Hypothesis: The support level of Republican and Democrats on the water-project are the same.  Due to a p value of .929 we fail to reject the null hypothesis.**

In [0]:
## Try a stretch goal
## I will make a function that when given a string column name will return a 
## ttest of republican vs democrat groups for the given column
def two_sample_ttest(column):
  return(ttest_ind(reps[column], dems[column], nan_policy='omit'))


column_list = ['handicapped-infants','water-project',
                          'budget','physician-fee-freeze', 'el-salvador-aid',
                          'religious-groups','anti-satellite-ban',
                          'aid-to-contras','mx-missile','immigration',
                          'synfuels', 'education', 'right-to-sue','crime','duty-free',
                          'south-africa']

In [35]:
## check to see if the code I did worked using the data I worked on earlier

two_sample_ttest('water-project')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)